clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3674,14 +4061,261 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool handleNonDecompStruct(const CXXRecordDecl *RD,
+                             const CXXBaseSpecifier &BS, QualType Ty) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool handleScalarType(FieldDecl *FD, QualType FieldTy) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool handleScalarType(ParmVarDecl *, QualType) final {
+    Expr *ParamRef = createParamReferenceExpr();
+    ArgExprs.push_back(ParamRef);
+    return true;
+  }
+
+  bool handleUnionType(FieldDecl *FD, QualType FieldTy) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool handleUnionType(ParmVarDecl *, QualType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool enterStruct(const CXXRecordDecl *RD, FieldDecl *FD, QualType Ty) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool enterStruct(const CXXRecordDecl *, ParmVarDecl *, QualType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool leaveStruct(const CXXRecordDecl *, FieldDecl *FD, QualType Ty) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool leaveStruct(const CXXRecordDecl *, ParmVarDecl *, QualType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool enterStruct(const CXXRecordDecl *RD, const CXXBaseSpecifier &BS,
+                   QualType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool leaveStruct(const CXXRecordDecl *RD, const CXXBaseSpecifier &BS,
+                   QualType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool enterArray(FieldDecl *FD, QualType ArrayType,
+                  QualType ElementType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool enterArray(ParmVarDecl *PD, QualType ArrayType,
+                  QualType ElementType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool leaveArray(FieldDecl *FD, QualType ArrayType,
+                  QualType ElementType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
+
+  bool leaveArray(ParmVarDecl *PD, QualType ArrayType,
+                  QualType ElementType) final {
+    // TODO
+    unsupportedFreeFunctionParamType();
+    return true;
+  }
 };
 
 // Kernels are only the unnamed-lambda feature if the feature is enabled, AND
 // the first template argument has been corrected by the library to match the
 // functor type.
 static bool IsSYCLUnnamedKernel(SemaSYCL &SemaSYCLRef, const FunctionDecl *FD) {
+  // If free function then remaining checks are not applicable.
+  if (isFreeFunction(SemaSYCLRef, FD))
+    return false;
```

Comment:
In C++23, following the adoption of [P1169R4 (static `operator()`)](https://wg21.link/P1169R4), the call operator of a closure type may be a static member and could be considered eligible for use as a SYCL free function kernel. Clang implemented support for such lambdas in Clang 16. We should add a test and either make sure an appropriate diagnostic is issued or that such functions work as expected.

Reply:
Currently we only support free functions at file scope.

Reply:
What happens if we use it now?